### DISCUSSION TITLE
[feature request] Configurable user frame for ABB work object frames

### DISCUSSION THREAD

**Discussion by kevingessner** [visose/Robots#271] [Q&A] [answered]
On ABB*, a work object frame is defined as two coordinate transforms: the user frame first, then the work object frame relative to the user frame. Currently, the Robots 'Create Frame' puts the specified frame as the user frame, and a zero frame for the work object frame: 

```
TASK PERS wobjdata Frame000:=[FALSE,TRUE,"",[[-200,400,0],[1,0,0,0]],[[0,0,0],[1,0,0,0]]];
TASK PERS wobjdata Frame001:=[FALSE,TRUE,"",[[-150,400,0],[1,0,0,0]],[[0,0,0],[1,0,0,0]]];
```

It would be useful to have the work object frames specified relative to a single user frame declared as its own variable, e.g.: 

```
TASK PERS pose FrameUserBase:=[[0,0,0],[1,0,0,0]];
TASK PERS wobjdata Frame000:=[FALSE,TRUE,"",FrameUserBase,[[-200,400,0],[1,0,0,0]]];
TASK PERS wobjdata Frame001:=[FALSE,TRUE,"",FrameUserBase,[[-150,400,0],[1,0,0,0]]];
```

With the default FrameUserBase, the resulting work object frames would be the same as before. The new variable would enable easily repositioning all of the work objects by changing the FrameUserBase.

Ideally, the user frame would be specified as a parameter (to set the default value), but even a hard-coded zero default that could be changed on the robot at runtime would be useful.

\* maybe on other robots too? #182 sounds similar for UR robots.

**Comment by visose** [comment]
Thank you for the suggestion, I think is a good idea. I'll try to add it to the next release, should not take long to do.

**Comment by kevingessner** [comment]
@visose Cool! I can take a crack a PR, but it will just be for ABB/Rapid -- I'm not familiar with any of the other systems.

**Reply by visose** [reply]
Don't worry, I'm on it.

**Reply by kevingessner** [reply]
Unfortunately some testing shows that the syntax I was hoping to generate isn't valid Rapid:

```
TASK PERS pose FrameUserBase:=[[0,0,0],[1,0,0,0]];
! can't use a variable inside a declaration
TASK PERS wobjdata Frame000:=[FALSE,TRUE,"",FrameUserBase,[[-200,400,0],[1,0,0,0]]];
TASK PERS wobjdata Frame001:=[FALSE,TRUE,"",FrameUserBase,[[-150,400,0],[1,0,0,0]]];
```

Instead the `uframe` of each work object needs to be set procedurally, like:

```
TASK PERS pose FrameUserBase:=[[0,0,0],[1,0,0,0]];
TASK PERS wobjdata Frame000:=[FALSE,TRUE,"",[[0,0,0],[1,0,0,0]],[[-200,400,0],[1,0,0,0]]];
TASK PERS wobjdata Frame001:=[FALSE,TRUE,"",[[0,0,0],[1,0,0,0]],[[-150,400,0],[1,0,0,0]]];

! In Proc Main:
Frame000.uframe := FrameUserBase;
Frame001.uframe := FrameUserBase;
```

It's admittedly kinda ugly, and I'm not sure Robots needs to support that pattern out of the box. At the very least, it would be great to have the `wobjdata` objects with the `Frame` as the final parameter, so the `uframe` can be safely changed.

**Reply by visose** [reply, marked as answer]
As you say, adding the extra declarations might be overkill for the case where you want to manually edit the user frame.

Another problem is that since currently Robots sets the user frame, changing it to set the object frame will be a breaking change. There might be someone tweaking the generated RAPID that relies on this. I might change it in the future with another set of breaking changes.

If the purpose is to make small adjustments to `FrameUserBase` and don't expect these changes to feed back to Grasshopper to re-run the simulation, one option is to use a custom command:
![image](https://github.com/visose/Robots/assets/5129614/fa2246ea-90d4-4266-a7d9-5b4dcbd780f8)

Notice that it swaps the user and object frames before assigning `FrameUserBase` to the user frame.

**Comment by kevingessner** [comment]
That make sense, very cool. I'm happy to close this then!
